File: crates/server/src/bench.rs
//! The measurement the server exposes.
//!
//! This runs the same `Pipeline` the native benchmarks run, over a capture the
//! `synth` crate generates, and reports the same statistics with the same
//! discipline — including refusing to quote a difference smaller than the
//! run-to-run spread.
//!
//! That last part matters more here than anywhere else in the project. The
//! server is expected to run on shared, burstable cloud CPU, where the spread is
//! large. Reporting a clean number from such a host would be the one dishonest
//! thing in this repository, so the spread travels with every figure and a
//! difference inside it is labelled unresolved.

use std::time::Instant;

use book::{BTreeLevels, Levels, SortedLevels};
use deep::Symbol;
use pipeline::{Histogram, Pipeline, Stage};

/// Which level container to run. The bake-off, exposed as a knob.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Container {
    BTree,
    Sorted,
}

impl Container {
    pub fn parse(s: &str) -> Self {
        match s {
            "btree" => Container::BTree,
            _ => Container::Sorted,
        }
    }
    pub fn name(self) -> &'static str {
        match self {
            Container::BTree => "BTreeMap (R0)",
            Container::Sorted => "sorted SoA (R1)",
        }
    }
}

/// Knobs a visitor can turn. All bounded: the server never takes code, a path,
/// or an unbounded size.
#[derive(Debug, Clone, Copy)]
pub struct Request {
    pub container: Container,
    /// Fraction of segments sent twice, modelling the B feed.
    pub duplicate_rate: f64,
    /// Fraction of segments dropped, modelling loss.
    pub gap_rate: f64,
    /// Offered load as a fraction of measured capacity, for the
    /// coordinated-omission demonstration.
    pub load: f64,
}

impl Request {
    pub fn clamped(mut self) -> Self {
        self.duplicate_rate = self.duplicate_rate.clamp(0.0, 1.0);
        self.gap_rate = self.gap_rate.clamp(0.0, 0.25);
        self.load = self.load.clamp(0.1, 2.0);
        self
    }
}

#[derive(Debug, Clone)]
pub struct TopOfBook {
    pub symbol: String,
    pub bid: Option<(String, u32)>,
    pub ask: Option<(String, u32)>,
    pub bid_levels: usize,
    pub ask_levels: usize,
}

#[derive(Debug, Clone)]
pub struct Report {
    pub container: &'static str,
    pub packets: u64,
    pub messages: u64,
    pub book_updates: u64,
    pub orders: u64,
    /// Best of the repetitions, nanoseconds per packet.
    pub ns_per_packet: f64,
    /// Worst minus best. Every figure here is read against this.
    pub spread: f64,
    pub duplicates_suppressed: u64,
    pub gaps_detected: u64,
    pub gaps_injected: u64,
    pub crossed_when_stable: u64,
    pub service_p50: u64,
    pub service_p99: u64,
    pub corrected_p50: u64,
    pub corrected_p99: u64,
    pub behind_schedule: u64,
    pub top: Vec<TopOfBook>,
}

const REPS: usize = 5;

/// Capture held in memory, generated once and reused across requests.
pub struct Corpus {
    pub frames: Vec<Vec<u8>>,
    pub gaps_injected: u64,
}

pub fn build_corpus(req: Request, segments: usize) -> Corpus {
    let cfg = synth::Config {
        segments,
        duplicate_rate: req.duplicate_rate,
        gap_rate: req.gap_rate,
        ..Default::default()
    };
    let cap = synth::generate(&cfg);
    // Split into frames once, so the timed loop measures the pipeline rather
    // than the capture reader.
    let mut frames = Vec::new();
    for pkt in iextp::capture::Capture::open(&cap.bytes).expect("generated capture is readable") {
        frames.push(pkt.data.to_vec());
    }
    Corpus {
        frames,
        gaps_injected: cap.gaps,
    }
}

pub fn run(req: Request, corpus: &Corpus) -> Report {
    match req.container {
        Container::BTree => run_with::<BTreeLevels>(req, corpus),
        Container::Sorted => run_with::<SortedLevels>(req, corpus),
    }
}

fn run_with<L: Levels>(req: Request, corpus: &Corpus) -> Report {
    let frames = &corpus.frames;

    // Warm up before timing anything. The first pass pays every page fault for
    // the symbol table and level storage; timing it reports the allocator, and
    // in this project it once produced an impossible negative stage cost.
    {
        let mut p: Pipeline<L> = Pipeline::with_capacity(1024);
        for f in frames {
            p.on_frame_fast(f);
        }
        std::hint::black_box(&p.counters);
    }

    let mut best = f64::MAX;
    let mut worst = 0f64;
    let mut counters = Default::default();
    let mut seq_stats = iextp::sequencer::Stats::default();
    let mut crossed = 0u64;
    let mut top = Vec::new();

    for _ in 0..REPS {
        let mut p: Pipeline<L> = Pipeline::with_capacity(1024);
        let start = Instant::now();
        for f in frames {
            p.on_frame_fast(f);
        }
        let ns = start.elapsed().as_nanos() as f64 / frames.len().max(1) as f64;
        if ns < best {
            best = ns;
        }
        if ns > worst {
            worst = ns;
        }
        counters = p.counters;
        seq_stats = p.seq.stats;
        crossed = p.book.stats.crossed_when_stable;
        top = snapshot(&p);
    }

    // Coordinated omission, at the requested fraction of measured capacity.
    let capacity = 1e9 / best.max(1.0);
    let rate = capacity * req.load;
    let per_item = 1e9 / rate;
    let mut service = Histogram::new();
    let mut corrected = Histogram::new();
    let mut behind = 0u64;
    {
        let mut p: Pipeline<L> = Pipeline::with_capacity(1024);
        let origin = Instant::now();
        let mut queue_free_at = 0u64;
        for (i, f) in frames.iter().enumerate() {
            let scheduled = (i as f64 * per_item) as u64;
            let work_start = origin.elapsed().as_nanos() as u64;
            p.on_frame_fast(f);
            let work_end = origin.elapsed().as_nanos() as u64;
            let took = work_end.saturating_sub(work_start);
            let started = queue_free_at.max(scheduled);
            let finished = started + took;
            queue_free_at = finished;
            if started > scheduled {
                behind += 1;
            }
            service.record(took);
            corrected.record(finished.saturating_sub(scheduled));
        }
    }

    Report {
        container: req.container.name(),
        packets: counters.packets,
        messages: counters.messages,
        book_updates: counters.book_updates,
        orders: counters.orders_encoded,
        ns_per_packet: best,
        spread: worst - best,
        duplicates_suppressed: seq_stats.duplicates,
        gaps_detected: seq_stats.gaps_opened,
        gaps_injected: corpus.gaps_injected,
        crossed_when_stable: crossed,
        service_p50: service.quantile(0.50),
        service_p99: service.quantile(0.99),
        corrected_p50: corrected.quantile(0.50),
        corrected_p99: corrected.quantile(0.99),
        behind_schedule: behind,
        top,
    }
}

/// One stage's measured cost.
pub struct StageCost {
    pub name: &'static str,
    pub cumulative: f64,
    pub this_stage: f64,
    pub spread: f64,
    /// True when the stage's cost is smaller than its own run-to-run spread,
    /// in which case it is not a measurement and must not be quoted as one.
    pub unresolved: bool,
}

/// The stage budget, by cumulative prefix.
///
/// Could not be resolved on the development laptop: five of six stages came in
/// under their own noise floor, because macOS offers no CPU pinning and the M2
/// mixes performance and efficiency cores. A dedicated, idle VM may do better --
/// which is the point of running it here rather than renting a machine to find
/// out.
///
/// Repetitions are round-robin across stages, never blocked per stage. Timing
/// all runs of one stage before the next lets load drift accumulate along the
/// stage order, and differencing then produces a negative cost. That is not
/// hypothetical; it happened, and a negative time is the only reason it was
/// noticed.
pub fn stage_budget(corpus: &Corpus) -> Vec<StageCost> {
    const REPS: usize = 9;
    let frames = &corpus.frames;

    // Warm on the FULL pipeline first, or the first prefix to build a book pays
    // every page fault and the stage after it comes out negative.
    {
        let mut p: Pipeline<SortedLevels> = Pipeline::with_capacity(1024);
        for f in frames {
            p.on_frame(f, Stage::Encode);
        }
        std::hint::black_box(&p.counters);
    }

    let mut best = vec![f64::MAX; Stage::ALL.len()];
    let mut worst = vec![0f64; Stage::ALL.len()];
    for _ in 0..REPS {
        for (i, stage) in Stage::ALL.iter().enumerate() {
            let mut p: Pipeline<SortedLevels> = Pipeline::with_capacity(1024);
            let start = Instant::now();
            for f in frames {
                p.on_frame(f, *stage);
            }
            let ns = start.elapsed().as_nanos() as f64 / frames.len().max(1) as f64;
            best[i] = best[i].min(ns);
            worst[i] = worst[i].max(ns);
        }
    }

    let mut out = Vec::new();
    let mut last = 0f64;
    for (i, stage) in Stage::ALL.iter().enumerate() {
        let delta = best[i] - last;
        let spread = worst[i] - best[i];
        out.push(StageCost {
            name: stage.name(),
            cumulative: best[i],
            this_stage: delta,
            spread,
            unresolved: delta < spread,
        });
        last = best[i];
    }
    out
}

/// Top of book for a handful of symbols, so a visitor sees an actual market
/// rather than only statistics.
fn snapshot<L: Levels>(p: &Pipeline<L>) -> Vec<TopOfBook> {
    let mut out = Vec::new();
    for t in ["AAPL", "MSFT", "SPY", "QQQ", "NVDA", "TSLA"] {
        let sym = Symbol::from_ticker(t);
        if let Some(sb) = p.book.get(sym) {
            out.push(TopOfBook {
                symbol: t.to_string(),
                bid: sb.best_bid().map(|(p, s)| (p.to_string(), s)),
                ask: sb.best_ask().map(|(p, s)| (p.to_string(), s)),
                bid_levels: sb.bids.len(),
                ask_levels: sb.asks.len(),
            });
        }
    }
    out
}

File: cpp/src/bench.cpp
// The C++ side of the cross-language comparison.
//
// Reports the same figures as the Rust `budget` example, over the same capture,
// with the same discipline: warm before timing, round-robin repetitions, and a
// run-to-run spread printed beside every number so a difference smaller than the
// noise is not read as a result.
//
//     ./ll_bench <capture.pcap> [reps]
#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <string>
#include <vector>

#include "nb/book.hpp"
#include "nb/capture.hpp"
#include "nb/deep.hpp"

namespace {

std::vector<std::uint8_t> read_file(const char* path) {
    std::ifstream f(path, std::ios::binary | std::ios::ate);
    if (!f) return {};
    const auto n = static_cast<std::size_t>(f.tellg());
    f.seekg(0);
    std::vector<std::uint8_t> buf(n);
    f.read(reinterpret_cast<char*>(buf.data()), static_cast<std::streamsize>(n));
    return buf;
}

struct Counters {
    std::uint64_t packets = 0;
    std::uint64_t messages = 0;
    std::uint64_t book_updates = 0;
    std::uint64_t crossed = 0;
    std::size_t levels = 0;
    std::uint64_t total_size = 0;
};

/// One full pass: pcap record -> frame -> segment -> DEEP -> book.
Counters run_once(const std::vector<nb::Bytes>& frames) {
    nb::Book book(8192);
    Counters c;
    for (const auto& f : frames) {
        ++c.packets;
        const auto dg = nb::parse_frame(f);
        if (!dg) continue;
        const auto seg = nb::parse_segment(dg->payload);
        if (!seg) continue;
        nb::MessageIter it(*seg);
        while (const auto body = it.next()) {
            const auto msg = nb::parse_message(*body);
            if (!msg) continue;
            ++c.messages;
            if (const auto* u = std::get_if<nb::PriceLevelUpdate>(&*msg)) {
                book.apply(*u);
                ++c.book_updates;
            }
        }
    }
    c.crossed = book.stats.crossed_when_stable;
    c.levels = book.total_levels();
    c.total_size = book.total_size();
    return c;
}

}  // namespace

int main(int argc, char** argv) {
    if (argc < 2) {
        std::fprintf(stderr, "usage: ll_bench <capture.pcap> [reps]\n");
        return 2;
    }
    const int reps = argc > 2 ? std::atoi(argv[2]) : 7;

    const auto buf = read_file(argv[1]);
    if (buf.empty()) {
        std::fprintf(stderr, "%s: cannot read\n", argv[1]);
        return 1;
    }

    // Collect frames up front so the timed loop measures the pipeline, not the
    // capture reader.
    std::vector<nb::Bytes> frames;
    {
        auto r = nb::PcapReader::open(nb::Bytes(buf));
        if (!r) {
            std::fprintf(stderr, "%s: not a pcap file\n", argv[1]);
            return 1;
        }
        while (const auto p = r->next()) frames.push_back(p->data);
    }
    std::printf("%s\n  %zu packets\n\n", argv[1], frames.size());

    // Warm before timing. The first pass faults in every page of the symbol
    // table and level storage; timing it measures the allocator.
    const Counters warm = run_once(frames);

    double best = 1e18, worst = 0.0;
    for (int i = 0; i < reps; ++i) {
        const auto t0 = std::chrono::steady_clock::now();
        const Counters c = run_once(frames);
        const auto t1 = std::chrono::steady_clock::now();
        // Keep the result observable so the optimiser cannot delete the work.
        asm volatile("" : : "r"(&c) : "memory");
        const double ns =
            std::chrono::duration<double, std::nano>(t1 - t0).count() /
            static_cast<double>(frames.size());
        if (ns < best) best = ns;
        if (ns > worst) worst = ns;
    }

    std::printf("  %-26s %10.2f ns/packet\n", "tick-to-book (C++)", best);
    std::printf("  %-26s %10.2f ns   run-to-run\n", "spread", worst - best);
    std::printf("  %-26s %10.2f ns/message\n", "per message",
                best * static_cast<double>(frames.size()) /
                    static_cast<double>(warm.messages ? warm.messages : 1));
    std::printf("\n  messages       %12llu\n", (unsigned long long)warm.messages);
    std::printf("  book updates   %12llu\n", (unsigned long long)warm.book_updates);
    std::printf("  levels held    %12zu\n", warm.levels);
    std::printf("  total size     %12llu\n", (unsigned long long)warm.total_size);
    std::printf("  crossed@stable %12llu  %s\n", (unsigned long long)warm.crossed,
                warm.crossed == 0 ? "<- clean" : "<- BOOK CORRUPTION");
    return 0;
}
